### ISSUE TITLE
auth0-sdk.js PrivateRoute redirect issue

### ISSUE THREAD

**Issue by davidkutas** [PedroBern/react-tiger-transition#5]
Hey,

We are working with a auth0-sdk.js (https://auth0.com/docs/quickstart/spa/react/01-login) and your librabry. We would like to bring them together.

The issue we are facing is how you declare the <Navigation> compared to <Switch>. When we use react-router-dom Switch all is good, routes, private routes can be declared properly! When change <Switch> to <Navigation> the app always want to redirect me back to the login page even if I started with a <Route> which is a public route. Thing is, the auth0-sdk.js PrivateRoute component detects the PrivateRoute even if the path in the browser is a public route.

**Comment by PedroBern** [comment]
Hi, please provide some code. Maybe I can help if you show the code both for the react-router-dom (that works )and for this package (that doesn't work).

**Comment by davidkutas** [comment]
hi, sure!

Baiscally I was able to recreate the problem by doing one of your examples:

`<Router>
    {/* BrowserRouter from react-router-dom */}

    {/* Context provider for transitions */}
    <Navigation>
      {/* Use Route the same way you use
            react-router Route with children */}
      <Route exact path="/">
        {/* Screen is just a div container
              with some basic style */}
        <Screen
          style={{
            backgroundColor: '#ffa000',
            ...screenStyle,
          }}
        >
          {/* Use Link the same way you use
                react-router Link, but
                add transition */}
          <Link to="/a" transition="glide-left">
            Check out the page A
          </Link>
        </Screen>
      </Route>

      <PrivateRoute
        exact
        path="/a"
        screen // shorthand to wrap children with screen
        screenProps={{
          style: {
            backgroundColor: '#607d8b',
            ...screenStyle,
          },
        }}
      >
        <Link to="/" transition="glide-right">
          Back to home page
        </Link>
      </PrivateRoute>
    </Navigation>
  </Router>`

If you see in your example I changed the second route to be a PrivateRoute. This idea comes from auth0-spa.js:

```
import React, { useEffect } from "react";
import PropTypes from "prop-types";
import { Route, withRouter } from "react-router-dom";
import { useAuth0 } from "../react-auth0-spa";

const PrivateRoute = ({ component: Component, path, ...rest }) => {
  const { loading, isAuthenticated, loginWithRedirect } = useAuth0();

  useEffect(() => {
    if (loading || isAuthenticated) {
      return;
    }
    const fn = async () => {
      await loginWithRedirect({
        appState: { targetUrl: path }
      });
    };
    fn();
  }, [loading, isAuthenticated, loginWithRedirect, path]);

  const render = props =>
    isAuthenticated === true ? <Component {...props} /> : null;

  return <Route path={path} render={render} {...rest} />;
};

PrivateRoute.propTypes = {
  component: PropTypes.oneOfType([PropTypes.element, PropTypes.func])
    .isRequired,
  path: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string)
  ]).isRequired
};

export default withRouter(PrivateRoute);
```

As soon as I do this (the PrivateRoute) even though the path is localhost:3000/ the PrivateRoute  is checking the path and the path parameter shows /a even though we are on localhost:3000/ 

![Screen Shot 2020-05-19 at 8 13 02 AM](https://user-images.githubusercontent.com/19826357/82344314-c1035d00-99a8-11ea-9b3a-2b5e0d57975f.png)

This PrivateRoutechecker comes straight from Auth0: https://github.com/auth0-samples/auth0-react-samples/blob/master/01-Login/src/components/PrivateRoute.js

Basically, with <Switch> my code is working, whenever I change to <Navigation> that's when funky stuff starts to appear.

**Comment by davidkutas** [comment]
We figured it out. It was a mistake on our side. The PrivateRoute component was messed up; the good one is here for the future:

```
import React, { useEffect } from 'react';
import { Route } from 'react-router-dom';
import { useAuth0 } from '../../../authentication/auth0-spa';

// Takes another component as one of its arguments.
// Using the useEffect hook it redirects to the login page if not yet authenticated.
// Components requiring authentication can be protected by wrapping into <PrivateRoute> component.
const PrivateRoute = ({ component: Component, path, ...rest }) => {
  const { loading, isAuthenticated, loginWithRedirect } = useAuth0();

  useEffect(() => {
    if (loading || isAuthenticated) {
      return;
    }
    const fn = async () => {
      await loginWithRedirect({
        appState: { targetUrl: window.location.pathname },
      });
    };
    fn();
  }, [loading, isAuthenticated, loginWithRedirect, path]);

  // eslint-disable-next-line react/jsx-props-no-spreading
  const render = (props) => (isAuthenticated === true ? <Component {...props} /> : null);

  // eslint-disable-next-line react/jsx-props-no-spreading
  return <Route path={path} render={render} {...rest} />;
};

export default PrivateRoute;

```

**Comment by davidkutas** [comment]
We figured it out. It was a mistake on our side. The PrivateRoute component was messed up; the good one is here for the future:

```
import React, { useEffect } from 'react';
import { Route } from 'react-router-dom';
import { useAuth0 } from '../../../authentication/auth0-spa';

// Takes another component as one of its arguments.
// Using the useEffect hook it redirects to the login page if not yet authenticated.
// Components requiring authentication can be protected by wrapping into <PrivateRoute> component.
const PrivateRoute = ({ component: Component, path, ...rest }) => {
  const { loading, isAuthenticated, loginWithRedirect } = useAuth0();

  useEffect(() => {
    if (loading || isAuthenticated) {
      return;
    }
    const fn = async () => {
      await loginWithRedirect({
        appState: { targetUrl: window.location.pathname },
      });
    };
    fn();
  }, [loading, isAuthenticated, loginWithRedirect, path]);

  // eslint-disable-next-line react/jsx-props-no-spreading
  const render = (props) => (isAuthenticated === true ? <Component {...props} /> : null);

  // eslint-disable-next-line react/jsx-props-no-spreading
  return <Route path={path} render={render} {...rest} />;
};

export default PrivateRoute;

```

**Comment by davidkutas** [comment]
Basically, previously we thought that by fixing the problem we had with the PrivateRoute will fix the issue, but unfortunately no. Navigation/Route from tiger basically checks the PrivateRoute even if we are on a completely different path such as `localhost:3000/`

**Comment by PedroBern** [comment]
I didn't replicate it, I will try this as soon as I have time. I'm not sure right now, but maybe the mix of `<Route />` from `react-router-dom` inside the `<PrivateRoute />` and the public `<TigerRoutes/>`  is messing up.

Try something like this inside the `PrivateRoute`:

```diff
--- import { Route } from 'react-router-dom';
+++ import { Route } from "react-tiger-transition";
...

--- const render = (props) => (isAuthenticated === true ? <Component {...props} /> : null);
--- return <Route path={path} render={render} {...rest} />;

+++ // use children instead of render
+++ return <Route path={path} {...rest}>{/* your thing here */}</Route>
```

Then tell me if you make some progress.

**Comment by davidkutas** [comment]
Thanks for the feedback. We were exactly doing the same thing. We think that the problem is in the App.js

The thing that we observed:

```
<Navigation>
<route exact path="/"> children </>
<privateroute path="/workspaces"> children </>
<privateroute path="/profile"> children </>
</Navigation>
```
Whatever we pass into routes or privateroutes the Navigation already took in the first privateroute (workspaces) and goes to the privateroute component instead of landing on the public route (/)

**Comment by davidkutas** [comment]
The actual code we are at: 

```
import React from 'react';
import { BrowserRouter as Router, Switch } from 'react-router-dom';

// you need to import styles!
// one of:
// styles/main.css
// styles/main.min.css
import 'react-tiger-transition/styles/main.min.css';

import {
  Navigation, Screen, Route, Link, glide,
} from 'react-tiger-transition';
import { useSelector } from 'react-redux';
import { CSSTransition } from 'react-transition-group';
import PrivateRoute from './views/components/private';

import Login from './views/main/login';
import Workspaces from './views/main/workspaces';
import Profile from './views/main/profile';
import Settings from './views/main/settings';


import { useAuth0 } from './authentication/auth0-spa';
import history from './utils/history';

// inject glide styles
glide({
  name: 'glide-left',
});
glide({
  name: 'glide-right',
  direction: 'right',
});

const screenStyle = {
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
};

// you will need to set the height of  <Navigation /> wrapper,
// in this case, it is the root node,
// better to do this on your stylesheet
document.getElementById('publicContainer').style.height = '100vh';

// Main function of the SPA, containing all the routes.
function App() {
  const globalState = useSelector((state) => state);
  const authenticate = useAuth0();


  console.log(authenticate);
  // console.log(authenticate);

  if (authenticate.loading) {
    return <div>Loading...</div>;
  }
  return (
    <>
      <Router>
     
        <Navigation>
  
          <PrivateRoute screen exact path="/settings">
     
            <Settings />
     

          </PrivateRoute>


          <Route screen exact path="/login">
   
            <Login />
 

          </Route>

          <PrivateRoute screen exact path="/profile">
      
            <Profile />
        

          </PrivateRoute>

          <Route screen exact path="/workspaces">
       
            <Workspaces />
         

          </Route>


        </Navigation>
      </Router>
    </>
  );
}

export default App;

```

**Comment by PedroBern** [comment]
> Whatever we pass into routes or privateroutes the Navigation already took in the first privateroute ...

Did you change the routes to be from `react-tiger-transition` instead of `react-router-dom`?

**Comment by davidkutas** [comment]
yes we did. Look:

```
import React, { useEffect } from 'react';
import {
  Route,
} from 'react-tiger-transition';
import { useAuth0 } from '../../../authentication/auth0-spa';

// Takes another component as one of its arguments.
// Using the useEffect hook it redirects to the login page if not yet authenticated.
// Components requiring authentication can be protected by wrapping into <PrivateRoute> component.
const PrivateRoute = ({ component: Component, path, ...rest }) => {
  const { loading, isAuthenticated, loginWithRedirect } = useAuth0();
  debugger;
  useEffect(() => {
    if (loading || isAuthenticated) {
      return;
    }
    const fn = async () => {
      await loginWithRedirect({
        appState: { targetUrl: window.location.pathname },
      });
    };
    fn();
  }, [loading, isAuthenticated, loginWithRedirect]);


  // eslint-disable-next-line react/jsx-props-no-spreading
  return (
    <Route screen path={path} exact {...rest}>

      stuff


    </Route>
  );
};

export default PrivateRoute;
```

**Comment by davidkutas** [comment]
Just realized what `<Navigation>` does compared to `<Switch>`. Your `<Navigation>` module is inclusively rendering all routes for the first shot, then based on the path it chooses one. Compared to this `<Switch>` from `react-router-dom` is choosing to render a matching route exclusively (https://reacttraining.com/react-router/web/api/Switch). Is this correct what I think how the `<Navigation>` is working?

**Comment by PedroBern** [comment]
> Just realized what <Navigation> does compared to <Switch>. Your <Navigation> module is inclusively checking all routes for the first shot, then based on the path it renders one. Compared to this <Switch> from react-router-dom is choosing to render a matching route exclusively for the first shot the app loads (https://reacttraining.com/react-router/web/api/Switch). Is this correct what I think how the <Navigation> is working?

It does check all routes, [here](https://github.com/PedroBern/react-tiger-transition/blob/master/src/Navigation.js#L71) but it is used only for the `defaultRoute`. The other routes are just like react-router routes with children (neither render or component props).

Compared to the Switch, it does render all routes instead of only one. But only the one with the correct path will receive the `in` prop of the animation, so only one will be visible.

**Comment by davidkutas** [comment]
I see. Thanks. the problem that I see with this approach that it makes PrivateRoutes impossible to pull off with the logic private routes are usually done. What happens in normal `react-router-dom` and `<Switch>` circumstances is that is that once the path is pointing to the private route, the route wrapper checks if we are logged in or not, and if not then it takes us back to the homepage, and if yes, we can continue to the logged in page. But with your component's logic it is not possible, every time you have a private route it is going to go into that component and re-route you to the homepage or authentication page. For this, I think it would be wiser to apply the same logic as the `<Switch>` does. Would that be possible to do?

**Comment by PedroBern** [comment]
> I see. Thanks. the problem that I see with this approach that it makes PrivateRoutes impossible to pull off with the logic private routes are usually done. 

When I have free time I will be looking into it.

> For this, I think it would be wiser to apply the same logic as the <Switch> does. Would that be possible to do in a future release?

It is not possible, because the Switch renders only one component at a time. In order to have the transitions, all routes must be mounted. When some route receives the `In`, it will transition. If we still need to mount the next route, the transition will not work, unfortunately.

**Comment by davidkutas** [comment]
Well, after understanding the problem, we ended up writing our own version. If you are interested get in touch with me at kutasdavid@gmail.com, I'd be happy to contribute and promote your tiger.

**Comment by PedroBern** [comment]
Ok :)

**Comment by TotallWAR** [comment]
Hi!
I have the same problem.
I think algorithm can be next:
- check if current matched route NOT private
- if it's not private -> filter routes and remove private routes
- if it's private route -> remove others routes and this private component will execute its logic with redirect

Besides, i use react-router-config
